apps/s_client.c

```diff
@@ -3827,18 +3827,15 @@ static int user_data_process(struct user_data_st *user_data, size_t *len,
         char *cmd_start = buf_start;
 
         cmd_start[outlen] = '\0';
-        do {
-            cmd_start = strstr(cmd_start, "{");
-            if (cmd_start == buf_start && *(cmd_start + 1) == '{') {
-                /* The "{" is escaped, so skip it */
-                cmd_start += 2;
-                buf_start++;
-                user_data->bufoff++;
-                user_data->buflen--;
-                outlen--;
-                continue;
-            }
-        } while(0);
+        cmd_start = strstr(cmd_start, "{");
+        if (cmd_start == buf_start && *(cmd_start + 1) == '{') {
+            /* The "{" is escaped, so skip it */
+            cmd_start += 2;
+            buf_start++;
+            user_data->bufoff++;
+            user_data->buflen--;
+            outlen--;
+        }
```

Comment:
I believe the intention with the original code was different:

I.e., the cycle was intended to be there with (1) as condition and there is a missing break after the if().

Reply:
I thought about this for quite a while and came to the conclusion that continuing the loop makes zero sense (without considerable rewriting).  I am fairly confident that the intention is to match a _leading_ `{` with the option to escape it using `{{` and to _never_ match an internal `{`.

The critical part is the comparison with `buf_start`.  After the first iteration, it is **impossible** for `cmd_start == buf_start` to ever be true.  Even ignoring that, continuing the loop and finding an internal `{{` would skip a number leading of characters which is clearly incorrect.

Reply:
Yes, after the first iteration it is impossible to again fall into the true condition block. But still the first `cmd_start = strstr(cmd_start, "{");` is evaluated which is actually what we want - otherwise you're going to output everything until the end of line even if there is actually a command later on the line. I.e., situation where you have `{{... {command}...`
would not be handled correctly with your patch.

Reply:
Alternative solution instead of keeping the loop:
```suggestion
        cmd_start = strchr(cmd_start, '{');
        if (cmd_start == buf_start && *(cmd_start + 1) == '{') {
            /* The "{" is escaped, so skip it */
            cmd_start = strchr(cmd_start + 2, '}');
            buf_start++;
            user_data->bufoff++;
            user_data->buflen--;
            outlen--;
        }
```

Reply:
Immediately following is:
```c
       if (cmd_start == buf_start) {
            /* Command detected */
```
which means it only cares about commands at the start of the line which means the `{{... {command}...` case still isn't relevant.

Reply:
@t8m is correct about the original intention of the code. We want to match as much "text" as possible that doesn't include a command. If the buffer has "{{" at the start then the first iteration looks for it, and skips over the first "{". The second iteration then adds additional text all the way up to the next instance of "{", e.g. so "{{helloworld{cmd}" should match "{helloworld". If we don't loop we will match it in two chunks "{" and "helloworld" which is less than ideal when it gets passed to `SSL_write` because it will be sent in two records instead of one.

Reply:
My alternative fix in #20788.